File: Ch01/Ex1_12.scala
/**
  * Ex1_10.scala
  * Kevin C. Baird
  * SICP in Scala (http://www.scala-lang.org/documentation/)
  *
  * Output Pascal's Triangle via a recursive process
  * (Exercise 1.12)
  */

object Ex1_12 extends App {

  def pascalsTriangle(rowNum: Int): Array[Int] = {
    def sumPairs(pairs: Array[Int]): Array[Int] = {
      pairs.length match {
        case 0 => Array()
        case 1 => pairs
        case 2 => Array(pairs.slice(0,0) + pairs.slice(1,1))
        case 3 => Array(pairs.slice(0,0) + pairs.slice(1,1)) ++ sumPairs(pairs.slice(1, pairs.length))
      }
    }
    val prevRow  = pascalsTriangle(rowNum-1)
    val prevSums = sumPairs(prevRow)
    rowNum match {
      case 0 => Array()
      case 1 => Array(1)
      case 2 => Array(1,1)
      case _ => Array(1) ++ prevSums ++ Array(1)
    }
  }

}

/**
  * execute with
  * scala -i Ex1_12.scala -e "println(Ex1_12.pascalsTriangle(4))"
  * until I figure out a better way
  */

File: Ch01/Ex1_12.hs
{-
 - Ex1_12.hs
 - Kevin C. Baird
 - SICP in Haskell
 -
 - Output Pascal's Triangle via a recursive process
 - (Exercise 1.12)
 -
 -}

module Ex1_12 where

pascalsTriangle :: Int -> [Int]
pascalsTriangle rowNum
  | rowNum == 0 = []
  | rowNum == 1 = [1]
  | rowNum == 2 = [1,1]
  | otherwise   = [1] ++ prevSums ++ [1]
  where
    prevRow  = pascalsTriangle (rowNum-1)
    prevSums = sumPairs prevRow

sumPairs :: [Int] -> [Int]
sumPairs []       = []
sumPairs [n]      = [n]
sumPairs [x, y]   = [x+y]
sumPairs (x:y:zs) = (x+y) : sumPairs (y:zs)

